keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,594 @@
+- Refactor the kubeadm source structure to easily distinguish between
+frontend and backend. This work is out of scope for this proposal.
+- Use the "staging" pattern from k/k to gradually move kubeadm out.
+kubeadm already does not depend on k/k and this pattern was seen as redundant.
+
+## Proposal
+
+### Actions
+
+Once all implementation details are clear between kubeadm maintainers
+and stakeholder SIGs such as Release and Docs, the following actions should
+be taken in this exact order:
+1) Update the following tracking issue to collect all
+actions: https://github.com/kubernetes/kubeadm/issues/1949
+2) Implement tooling in the k/release repository to build kubeadm artifacts.
+3) Adapt the reference documentation tooling maintained by SIG Docs to build
+from the kubeadm repository.
+4) Implement a post-submit job that will sync the k/kubeadm repository
+tags and branches with those of k/k. Keep the job dry-running as a start.
+5) Announce the move on on the kubernetes-dev mailing list.
+6) Move kubeadm to the k/kubeadm repository in a "kubeadm" sub-module.
+7) Remove kubeadm and references to cmd/kubeadm from k/k.
+8) Enable the post-submit job.
+
+### User Stories
+
+#### Story 1
+
+As a user, I would like to build kubeadm without depending on the
+kubernetes/kubernetes repository.
+
+kubeadm no longer imports k/k packages, so this is already possible
+if one just copies cmd/kubeadm outside of k/k and creates a new Go
+module for it. It still requires some hacks around applying a version in the
+kubeadm binary. Once kubeadm moves out both the build tooling and module
+will be available to users for easy building.
+
+#### Story 2
+
+As a user, I would like to import the kubeadm public, versioned API without
+importing kubernetes/kubernetes.
+
+Currently importing the k/k root module will make your build fail.
+The best solution for users is to fork the kubeadm API packages.
+Once kubeadm moves out, importing the "kubeadm" module will make
+it possible to easily import only the API portion.
+
+#### Story 3
+
+As a user, I would like to import utility packages from kubeadm.
+
+Similar to the above user story, importing kubeadm from k/k will fail.
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository under
+a "kubeadm" folder with a sub-module.
+
+This makes it possible to use `go get k8s.io/kubeadm/kubeadm` to obtain
+the kubeadm source code. It also provides nice separation with the
+existing contents of the k/kubeadm repository such as "kinder", "operator"
+and various docs and tests.
+
+One side effect is that these contents will now be tagged and branched with
+kubeadm, but this is actually a positive change.
+
+#### Handling older kubeadm versions
```

Comment:
good point. i need to outline this too. but the TL;DR is:
- we need a custom cherry pick script if we want one because the k/k one will not work.
- the cherry pick policy is the same as k/k

Reply:
added a couple of sentences about cherry picks.